lib/vm/src/cache.rs

```diff
@@ -78,30 +96,31 @@ where
 
         // pop from lru cache if present
         if let Some(cache) = &mut self.instances {
-            let val = cache.pop(&hash);
-            if let Some(inst) = val {
-                inst.leave_storage(Some(deps.storage));
-                return Ok(inst);
+            if let Some(cached_instance) = cache.pop(&hash) {
+                self.stats.hits_instance += 1;
+                return Ok(Instance::from_wasmer(cached_instance, deps));
             }
         }
 
         // try from the module cache
         let res = self.modules.load_with_backend(hash, backend());
         if let Ok(module) = res {
+            self.stats.hits_module += 1;
             return Instance::from_module(&module, deps);
         }
 
         // fall back to wasm cache (and re-compiling) - this is for backends that don't support serialization
         let wasm = self.load_wasm(id)?;
+        self.stats.misses += 1;
         Instance::from_code(&wasm, deps)
     }
 
     pub fn store_instance(&mut self, id: &[u8], instance: Instance<S, A>) -> Option<Extern<S, A>> {
         if let Some(cache) = &mut self.instances {
             let hash = WasmHash::generate(&id);
             let storage = instance.take_storage();
-            let api = instance.api; // copy it
-            cache.put(hash, instance);
+            let (wasmer_instance, api) = Instance::recycle(instance);
```

Comment:
How about having `Instance::recycle` return `{wasmer_instance, extern}` and collapse the `take_storage` line above and simply the return value?